### ISSUE TITLE
RUF055 ignores backslashes in the `repl` argument of `re.sub`

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#14757]
In Ruff 0.8.1, [`unnecessary-regular-expression` (RUF055)](https://docs.astral.sh/ruff/rules/unnecessary-regular-expression/) reports a false positive for `re.sub` when the replacement string contains a backslash.

```console
$ cat ruf055.py
import re
print(re.sub(" ", r"\n", "a b"))

$ python ruf055.py
a
b

$ ruff check --isolated --preview --select RUF055 ruf055.py --fix
Found 1 error (1 fixed, 0 remaining).

$ cat ruf055.py
import re
print("a b".replace(" ", "\\n"))

$ python ruf055.py
a\nb
```

This issue was mentioned in https://github.com/astral-sh/ruff/pull/14679#discussion_r1863802187 with the example of `\g<0>`, which was deemed acceptably implausible. Because a replacement string can also include character escapes, this issue is plausible.

Alternatively, this could be considered a true positive, in which case the fix needs to interpret escape sequences in the replacement string.

**Comment by MichaReiser** [comment]
Thanks. 

Okay, so the part that we overlooked from the specification is:

> if it is a string, any backslash escapes in it are processed. That is, \n is converted to a single newline character, \r is converted to a carriage return, and so forth.

**Comment by MichaReiser** [comment]
@AlexWaygood pointed out that this could be addressed by https://github.com/astral-sh/ruff/pull/14679

**Comment by AlexWaygood** [comment]
Thanks @dscorbett, as ever! You were quite right that #14679 would not have fixed this in its earlier form. I pushed some updates to it, so I _think_ it now handles everything correctly, though. Let us know if you see any other issues with the check!
